### PR TITLE
MB-8384: txdb investigation

### PR DESCRIPTION
**Pull request by ahobson**
## Description

This uses go-txdb to create a transaction that is rolled back at the end of each test. This should give some speedups

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

I tried to make each commit independent and as self contained as possible, so reviewing each commit in turn might be a good strategy.

## Setup

```sh
make server_test_setup
time make server_test_standalone
```

For example:
```
# on master
$ go test -count=1 ./pkg/handlers/internalapi/
ok  	github.com/transcom/mymove/pkg/handlers/internalapi	44.086s
# on this branch
$ go test -count=1 ./pkg/handlers/internalapi/
ok  	github.com/transcom/mymove/pkg/handlers/internalapi	26.561s
```

## Code Review Verification Steps

* [ ] If the change is risky, it has been tested in experimental before merging.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [investigate go-txdb to speed up test suite](MB-8384) for this change

## Screenshots

If this PR makes visible UI changes, an image of the finished UI can help reviewers and casual
observers understand the context of the changes. A before image is optional and
can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow instead of using multiple images. You may want to use GIPHY CAPTURE for this! 📸

_Please frame screenshots to show enough useful context but also highlight the affected regions._

### CODE REVIEW

**Comment by monfresh** [comment]
@ahobson I'm reviewing it this week!

**Comment by ahobson** [comment]
> @ahobson I'm reviewing it this week!

Please let me know how I can help make it easier to review. I could do a walkthrough via zoom if that would be helpful

**Comment by monfresh** [comment]
A walkthrough would be great. I put a meeting on your calendar for tomorrow. Feel free to move to a better time if that one doesn't work.

File: pkg/handlers/internalapi/mto_shipment_test.go [hunk 1]
```diff
@@ -683,22 +725,33 @@ func (suite *HandlerSuite) TestListMTOShipmentsHandler() {
...
+		firstCreatedShipment := subtestData.shipments[0]
+		nextCreatedShipment := subtestData.shipments[1]
+		if nextCreatedShipment.CreatedAt.Before(firstCreatedShipment.CreatedAt) {
+			firstCreatedShipment = subtestData.shipments[1]
+			nextCreatedShipment = subtestData.shipments[0]
+		}
+		_, err := time.Parse(time.RFC3339, okResponse.Payload[0].CreatedAt.String())
+		if err != nil {
+			suite.TestLogger().Fatal("unable to parse string time")
+		}
```
Comment (felipe-lee) [inline pkg/handlers/internalapi/mto_shipment_test.go:744]: This part can be removed. I'd deleted it in a previous pr because I changed how we tested what this bit was trying to test. It's essentially lines 746-751 now do what this was doing, without resorting to checking times.

Comment (ahobson) [reply pkg/handlers/internalapi/mto_shipment_test.go]: I believe I addressed this

File: pkg/testingsuite/pop_suite_txn.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	return nil, nil
+}
+
+// TransactionContextOptions returns the current pop.Transaction
+// Needed to implement the pop.store interface, but won't actually be called
+func (store *PopSuiteTxnStore) TransactionContextOptions(ctx context.Context, _ *sql.TxOptions) (*pop.Tx, error) {
+	return nil, nil
+}
+
+// WrapTxnPopConnection sets up the pop Connection for this test
```
Comment (monfresh) [inline pkg/testingsuite/pop_suite_txn.go:101]: Did you mean `openTxnPopConnection` in this comment to match the function name?

File: pkg/testingsuite/pop_suite_txn.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	"fmt"
+	"log"
+	"regexp"
+
+	"github.com/DATA-DOG/go-txdb"
+	"github.com/gobuffalo/pop/v5"
+	"github.com/jmoiron/sqlx"
+)
+
+// use a driver that sqlx knows uses a doller binding like postgres
```
Comment (monfresh) [inline pkg/testingsuite/pop_suite_txn.go:15]: doller -> dollar

File: pkg/parser/pricing/parse_test.go [hunk 4]
```diff
@@ -39,20 +32,17 @@ func TestPricingParserSuite(t *testing.T) {
...
 
-	hs.tableFromSliceCreator = dbtools.NewTableFromSliceCreator(hs.DB(), logger, true, false)
-
 	hs.xlsxFile, err = xlsx.OpenFile(hs.xlsxFilename)
 	if err != nil {
 		logger.Panic("could not open XLSX file", zap.Error(err))
 	}
 
 	suite.Run(t, hs)
-	hs.PopTestSuite.TearDown()
```
Comment (monfresh) [inline pkg/parser/pricing/parse_test.go:55]: Was this removal intentional?

File: pkg/services/ghcimport/ghc_rateengine_importer_test.go [hunk 5]
```diff
@@ -93,7 +93,6 @@ func TestGHCRateEngineImportSuite(t *testing.T) {
 	}
 
 	suite.Run(t, hs)
-	hs.PopTestSuite.TearDown()
```
Comment (monfresh) [inline pkg/services/ghcimport/ghc_rateengine_importer_test.go:96]: In this case, I think this removal makes sense because this package defines a `TearDownSuite` function that calls `suite.PopTestSuite.TearDown()`. Is that correct?

File: pkg/services/move_task_order/move_task_order_service_test.go [hunk 6]
```diff
@@ -12,15 +12,9 @@ type MoveTaskOrderServiceSuite struct {
...
-	suite.FatalNoError(err)
-}
-
 func TestMoveTaskOrderServiceSuite(t *testing.T) {
 	ts := &MoveTaskOrderServiceSuite{
-		testingsuite.NewPopTestSuite(testingsuite.CurrentPackage()),
+		testingsuite.NewPopTestSuite(testingsuite.CurrentPackage(), testingsuite.WithPerTestTransaction()),
 	}
 	suite.Run(t, ts)
-	ts.PopTestSuite.TearDown()
```
Comment (monfresh) [inline pkg/services/move_task_order/move_task_order_service_test.go:25]: Was this removal intentional?

**Review by monfresh — APPROVED** [review]
This is great! Thanks for the hard work on this one, and for saving us time now and in the future.

File: pkg/handlers/ghcapi/mto_shipment_test.go [hunk 7]
```diff
@@ -1744,30 +1875,42 @@ func (suite *HandlerSuite) TestCreateMTOShipmentHandler() {
 		suite.NotEmpty(typedResponse.Payload.InvalidFields)
 	})
 
-	suite.T().Run("POST failure - 422 - invalid input, missing pickup address", func(t *testing.T) {
-		fetcher := fetch.NewFetcher(builder)
-		creator := mtoshipment.NewMTOShipmentCreator(suite.DB(), builder, fetcher)
+	//
+	// TODO: this test panics with a null pointer exception. Duncan
+	// said he would look into a fix - ahobson 3 June 2021
```
Comment (monfresh) [inline pkg/handlers/ghcapi/mto_shipment_test.go:1880]: This is failing because the validation errors are nil. The handler is receiving `services.InvalidInputError` from the service, but without validation errors, and then it's assuming there are validation errors and calls `payloadForValidationError`, which in turn calls `NewValidationErrorsResponse` which tries to iterate through `range verrs.Keys()`, but `verrs` is nil.

Comment (felipe-lee) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1880]: There's a jira bug to fix this at some point: https://dp3.atlassian.net/browse/MB-8326

Comment (monfresh) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1880]: The reason this is happening is because by the time the payload has arrived at the service, the `shipment.PickupAddress` is `nil`, and so the service is returning `services.NewInvalidInputError(uuid.Nil, nil, nil, "PickupAddress is required to create an HHG or NTS type MTO shipment")`, with `nil` validation errors.

It looks like this handler is not following the same pattern we use in other handlers, which is to pass in `validate.NewErrors()` as the last argument of `payloadForValidationError`. However, that means that the `InvalidFields` will be empty, but this test is expecting them to NOT be empty, which doesn't make sense.

It seems like we should not be calling `NewInvalidInputError` without any validation errors if we expect the `InvalidFields` to be populated.

cc @duncan-truss

Comment (duncan-truss) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1880]: Yep you're right I think that's what I outlined here https://dp3.atlassian.net/browse/MB-8326

File: Makefile [hunk 8]
```diff
@@ -654,6 +657,11 @@ db_test_run: db_test_start db_test_create ## Run Test DB
 .PHONY: db_test_reset
 db_test_reset: db_test_destroy db_test_run ## Reset Test DB (destroy and run)
 
+.PHONY: db_test_truncate
+db_test_truncate:
+	@echo "Truncating ${DB_NAME_TEST} database..."
+	DB_PORT=$(DB_PORT_TEST) DB_NAME=$(DB_NAME_TEST) ./scripts/db-truncate
+
```
Comment (felipe-lee) [inline Makefile:664]: I just merged something that also added this `make` target, but doesn't do it as well. It also included a `db_dev_truncate`, which could be converted to use this same script with the appropriate env vars. Likely should be done in this pr though, otherwise we'll end up with duplicate targets.

File: pkg/handlers/ghcapi/mto_shipment_test.go (hunk 7 above)
Comment (monfresh) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1880]: Looking at the other APIs, they define a `ValidationError` and `ClientError` methods in `model_to_payload`, but when I tried doing the same for the GHC API, it failed because it's expecting a different type. The difference is that in the Swagger YAML definition of `ValidationError` for the GHC API, it's defining `additionalProperties` as `type: string`, whereas the others APIs expect an array to represent a list of errors for the invalid field.

**Review by felipe-lee — APPROVED** [review]
LGTM, thanks for all the work you did. Just a comment on the `Makefile` targets regarding some changes that I just merged into master

File: pkg/testingsuite/pop_suite_txn.go (hunk 2 above)
Comment (ahobson) [reply pkg/testingsuite/pop_suite_txn.go:101]: yep! fixing now

File: pkg/services/ghcimport/ghc_rateengine_importer_test.go (hunk 5 above)
Comment (ahobson) [reply pkg/services/ghcimport/ghc_rateengine_importer_test.go:96]: Yep, that's exactly right!

File: pkg/handlers/ghcapi/mto_shipment_test.go (hunk 7 above)
Comment (monfresh) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1880]: OK, I get it now. The easiest fix is to update the `payloadForValidationError` function so that it only includes the `InvalidFields` if validation errors are present:
```golang
func payloadForValidationError(title string, detail string, instance uuid.UUID, validationErrors *validate.Errors) *ghcmessages.ValidationError {
    payload := &ghcmessages.ValidationError{
        ClientError:   *payloadForClientError(title, detail, instance),
    }

    if validationErrors != nil {
        payload.InvalidFields = handlers.NewValidationErrorsResponse(validationErrors).Errors
    }

    return payload
}
```
And then this test won't panic anymore, but the last assertion needs to be removed because the InvalidFields is going to be empty in this case. The fact that it was not empty before was probably because the test was not doing what we thought it was.

Comment (monfresh) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1880]: I'll push this change up since I'm already in here.

File: Makefile (hunk 8 above)
Comment (ahobson) [reply Makefile:663]: Yep, converted!
